telemetry/derive/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+/// corresponding to the literal provided in quotes, with the second
+/// argument being `TOTAL_STR == "total"`. If the execution of the
+/// `Fn`'s body doesn't result in an [`Err`] variant, another metric
+/// with the same first argument and `SUCCESS_STR = "success"` is also
+/// incremented. Thus one can infer the number of rejected
+/// transactions based on this parameter. If necessary, this macro
+/// should be edited to record different [`Err`] variants as different
+/// rejections, so we could (in theory), record the number of
+/// transactions that got rejected because of
+/// e.g. `SignatureCondition` failure.
+///
+/// If you also want to track the execution time of the `isi`, you
+/// should prefix the quoted metric with the `+` symbol.
+///
+/// # Examples
+/// ```rust
+/// use iroha_core::wsv::{World, WorldStateView};
+/// use iroha_telemetry_derive::metrics;
+///
+/// #[metrics(+"test_query", "another_test_query_without_timing")]
+/// fn execute(wsv: &WorldStateView<World>) -> Result<(), ()> {
+///		Ok(())
+/// }
+/// ```
+#[proc_macro_error]
+#[proc_macro_attribute]
+#[allow(clippy::str_to_string)]
+pub fn metrics(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let ItemFn {
+        attrs,
+        vis,
+        sig,
+        block,
+    }: ItemFn = parse_macro_input!(item as ItemFn);
+
+    // This is a good sanity check. Possibly redundant.
+    if sig.ident != "execute" {
+        abort!(sig.ident, "Function should be an `impl execute`");
+    }
+    match sig.output.clone() {
+        syn::ReturnType::Default => abort!(
+            sig.output,
+            "`Fn` must return `Result`. Returns nothing instead. "
+        ),
+        #[allow(clippy::string_to_string)]
+        syn::ReturnType::Type(_, typ) => match *typ {
+            Type::Path(pth) => {
+                let Path { segments, .. } = pth.path;
+                #[allow(clippy::expect_used)]
+                let type_name = &segments.iter().last().expect("non-empty path").ident;
+                if *type_name != "Result" {
+                    abort!(
+                        type_name,
+                        format!("Should return `Result`. Found {}", type_name)
+                    );
+                }
+            }
+            _ => abort!(
+                typ,
+                "Should return `Result`. Non-path result type specification found"
+            ),
+        },
+    }
+    if sig.inputs.is_empty() {
+        abort!(
+            sig,
+            "Function must have at least one argument of type `WorldStateView`."
+        );
+    }
+    let _specs = parse_macro_input!(attr as MetricSpecs);
+    // Again this may seem fragile, but if we move the metrics from
+    // the `WorldStateView`, we'd need to refactor many things anyway
+    let _metric_arg_ident = arg_metrics(&sig.inputs)
+        .unwrap_or_else(|args| abort!(args, "At least one argument must be a `WorldStateView`."));
```

Comment:
would you move these two down so they are guarded by `metric-instrumentation`

Reply:
This checks in one go if there is an argument with type `WorldStateView` which I want to also be present if `metric-instrumentation` is off. The idea is, that it warns you about mistakes if you don't turn the feature on, so that you don't get a nasty surprise.